File: protocol-generator/java/src/main/java/org/entitycore/protocol/crypto/Ed.java
package org.entitycore.protocol.crypto;

import java.math.BigInteger;
import java.security.GeneralSecurityException;
import java.security.KeyFactory;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.Signature;
import java.security.interfaces.EdECPublicKey;
import java.security.spec.EdECPoint;
import java.security.spec.EdECPrivateKeySpec;
import java.security.spec.EdECPublicKeySpec;
import java.security.spec.NamedParameterSpec;

import org.entitycore.protocol.crypto.PeerId.Curve;

/**
 * EdDSA sign / verify via the JDK SunEC provider (JEP 339: Ed25519 + Ed448), and raw
 * public-key extraction. Zero runtime dependency; RFC-8032 deterministic.
 *
 * <p><b>Crypto sourcing decision (A-JAVA-002).</b> SIGN/VERIFY use SunEC for both
 * curves — the JDK closes the agility SIGNATURE bar natively, with no BouncyCastle
 * (the CORE build is BC-free; BC stays an opt-in cross-check only). RAW PUBLIC-KEY
 * derivation uses {@link EdKeyDerivation} (native RFC-8032 scalar multiply), because
 * SunEC exposes a public key only as an {@link EdECPoint} and has no seed→public API
 * — and the JDK ships no SHAKE256, which the Ed448 expansion needs. So the public-key
 * derivation is hand-rolled (SHA-512 / hand-rolled SHAKE256 + BigInteger curve math),
 * keeping the whole core dependency-free.
 */
public final class Ed {
    private Ed() { }

    private static NamedParameterSpec spec(Curve curve) {
        return (curve == Curve.ED25519) ? NamedParameterSpec.ED25519 : NamedParameterSpec.ED448;
    }

    private static String algo(Curve curve) {
        return (curve == Curve.ED25519) ? "Ed25519" : "Ed448";
    }

    private static int seedLen(Curve curve) {
        return (curve == Curve.ED25519) ? 32 : 57;
    }

    /** Build a {@link PrivateKey} from a raw RFC-8032 seed (32 B Ed25519 / 57 B Ed448).
     *  SunEC runs the RFC-8032 expansion internally. */
    public static PrivateKey privateKeyFromSeed(byte[] seed, Curve curve)
            throws EntityCryptoException {
        if (seed.length != seedLen(curve)) {
            throw new BadSeedException(curve + " seed must be " + seedLen(curve)
                    + " bytes, got " + seed.length);
        }
        try {
            KeyFactory kf = KeyFactory.getInstance(algo(curve));
            return kf.generatePrivate(new EdECPrivateKeySpec(spec(curve), seed.clone()));
        } catch (GeneralSecurityException e) {
            throw new EntityCryptoException("private key from seed failed: " + curve, e);
        }
    }

    /** Derive the RAW (RFC-8032 byte-string) public key from a seed — native, via
     *  {@link EdKeyDerivation}. */
    public static byte[] rawPublicKeyFromSeed(byte[] seed, Curve curve)
            throws EntityCryptoException {
        return EdKeyDerivation.rawPublicKey(seed, curve);
    }

    /** Extract the raw RFC-8032 public-key bytes from a SunEC {@link PublicKey}
     *  ({@link EdECPoint} → little-endian y with the x sign bit in the final byte). */
    public static byte[] rawPublicKey(PublicKey pub, Curve curve) throws EntityCryptoException {
        if (!(pub instanceof EdECPublicKey ed)) {
            throw new EntityCryptoException("not an EdEC public key: " + pub.getClass());
        }
        EdECPoint point = ed.getPoint();
        int len = seedLen(curve);
        byte[] out = new byte[len];
        byte[] y = point.getY().toByteArray(); // big-endian magnitude
        for (int i = 0; i < y.length; i++) {
            int idx = y.length - 1 - i;        // least-significant first
            if (i < len) {
                out[i] = y[idx];
            }
        }
        if (point.isXOdd()) {
            out[len - 1] |= (byte) 0x80;
        }
        return out;
    }

    /** Sign {@code message} with the seed-derived key. RFC-8032 signature (64 B
     *  Ed25519, 114 B Ed448). */
    public static byte[] sign(byte[] seed, byte[] message, Curve curve)
            throws EntityCryptoException {
        try {
            PrivateKey priv = privateKeyFromSeed(seed, curve);
            Signature sig = Signature.getInstance(algo(curve));
            sig.initSign(priv);
            sig.update(message);
            return sig.sign();
        } catch (GeneralSecurityException e) {
            throw new EntityCryptoException("sign failed: " + curve, e);
        }
    }

    /** Verify a signature against a RAW public key. */
    public static boolean verify(byte[] rawPublicKey, byte[] message, byte[] signature,
                                 Curve curve) throws EntityCryptoException {
        try {
            PublicKey pub = publicKeyFromRaw(rawPublicKey, curve);
            Signature sig = Signature.getInstance(algo(curve));
            sig.initVerify(pub);
            sig.update(message);
            return sig.verify(signature);
        } catch (GeneralSecurityException e) {
            return false;
        }
    }

    /** Rebuild a SunEC {@link PublicKey} from RAW RFC-8032 bytes (inverse of
     *  {@link #rawPublicKey}). */
    public static PublicKey publicKeyFromRaw(byte[] raw, Curve curve)
            throws EntityCryptoException {
        try {
            byte[] le = raw.clone();
            int last = le.length - 1;
            boolean xOdd = (le[last] & 0x80) != 0;
            le[last] &= 0x7f;
            byte[] be = new byte[le.length];
            for (int i = 0; i < le.length; i++) {
                be[i] = le[le.length - 1 - i];
            }
            BigInteger y = new BigInteger(1, be);
            EdECPoint point = new EdECPoint(xOdd, y);
            KeyFactory kf = KeyFactory.getInstance(algo(curve));
            return kf.generatePublic(new EdECPublicKeySpec(spec(curve), point));
        } catch (GeneralSecurityException e) {
            throw new EntityCryptoException("public key from raw failed: " + curve, e);
        }
    }
}

File: protocol-generator/kotlin/src/main/kotlin/org/entitycore/protocol/crypto/Ed.kt
package org.entitycore.protocol.crypto

import org.entitycore.protocol.EcfException
import org.entitycore.protocol.EntityError
import java.security.KeyFactory
import java.security.PrivateKey
import java.security.PublicKey
import java.security.Signature
import java.security.interfaces.EdECPublicKey
import java.security.spec.EdECPoint
import java.security.spec.EdECPrivateKeySpec
import java.security.spec.EdECPublicKeySpec
import java.security.spec.NamedParameterSpec
import java.math.BigInteger

/** The signature curves with allocated key_type codes (V7 §1.5). */
enum class Curve { ED25519, ED448 }

/**
 * EdDSA sign / verify via the JDK SunEC provider (JEP 339: Ed25519 + Ed448), and raw
 * public-key extraction. Zero runtime dependency; RFC-8032 deterministic. Called
 * directly from Kotlin (Kotlin/JVM has zero-overhead Java interop — `java.security`
 * is just a Kotlin import). Profile [codec].ed25519_library = jdk-sunec.
 *
 * The S2 floor needs only Ed25519 sign over canonical-ECF entity bytes (the
 * `signature.*` corpus); SunEC runs the RFC-8032 seed expansion internally, so a raw
 * 32-byte seed → deterministic 64-byte signature needs no third-party crypto. Ed448 +
 * the raw-key-from-point seam are wired for the (deferred, higher-bar) agility corpus.
 */
object Ed {

    private fun spec(curve: Curve): NamedParameterSpec =
        if (curve == Curve.ED25519) NamedParameterSpec.ED25519 else NamedParameterSpec.ED448

    private fun algo(curve: Curve): String = if (curve == Curve.ED25519) "Ed25519" else "Ed448"

    private fun seedLen(curve: Curve): Int = if (curve == Curve.ED25519) 32 else 57

    /** Build a [PrivateKey] from a raw RFC-8032 seed (32 B Ed25519 / 57 B Ed448).
     *  SunEC runs the RFC-8032 expansion internally. */
    fun privateKeyFromSeed(seed: ByteArray, curve: Curve): PrivateKey {
        if (seed.size != seedLen(curve)) {
            throw EcfException(EntityError.CryptoError.BadSeed(
                "$curve seed must be ${seedLen(curve)} bytes, got ${seed.size}"))
        }
        return KeyFactory.getInstance(algo(curve))
            .generatePrivate(EdECPrivateKeySpec(spec(curve), seed.copyOf()))
    }

    /** Sign [message] with the seed-derived key. RFC-8032 signature (64 B Ed25519,
     *  114 B Ed448). Deterministic. */
    fun sign(seed: ByteArray, message: ByteArray, curve: Curve): ByteArray {
        try {
            val priv = privateKeyFromSeed(seed, curve)
            val sig = Signature.getInstance(algo(curve))
            sig.initSign(priv)
            sig.update(message)
            return sig.sign()
        } catch (e: java.security.GeneralSecurityException) {
            throw EcfException(EntityError.CryptoError.SignFailed("sign failed: $curve: ${e.message}"))
        }
    }

    /**
     * Derive the raw RFC-8032 public-key bytes from a secret seed (S3 peer-id +
     * system/peer construction). SunEC has no seed→public-point API, so Ed25519 is
     * derived in [EdKeyDerivation] (A-KT-007). Only Ed25519 (the §9.1 floor curve) is
     * wired; Ed448 raw-key derivation is a deferred agility higher-bar.
     */
    fun rawPublicKeyFromSeed(seed: ByteArray, curve: Curve): ByteArray = when (curve) {
        Curve.ED25519 -> EdKeyDerivation.rawPublicKeyEd25519(seed)
        Curve.ED448 -> throw EcfException(EntityError.CryptoError.UnsupportedKeyType(
            "Ed448 seed→pubkey derivation is a deferred agility higher-bar"))
    }

    /** Verify a signature against a RAW public key. */
    fun verify(rawPublicKey: ByteArray, message: ByteArray, signature: ByteArray, curve: Curve): Boolean =
        try {
            val pub = publicKeyFromRaw(rawPublicKey, curve)
            val sig = Signature.getInstance(algo(curve))
            sig.initVerify(pub)
            sig.update(message)
            sig.verify(signature)
        } catch (e: java.security.GeneralSecurityException) {
            false
        }

    /** Extract the raw RFC-8032 public-key bytes from a SunEC [PublicKey]
     *  ([EdECPoint] → little-endian y with the x sign bit in the final byte). */
    fun rawPublicKey(pub: PublicKey, curve: Curve): ByteArray {
        val ed = pub as? EdECPublicKey
            ?: throw EcfException(EntityError.CryptoError.UnsupportedKeyType("not an EdEC public key: ${pub.javaClass}"))
        val point = ed.point
        val len = seedLen(curve)
        val out = ByteArray(len)
        val y = point.y.toByteArray() // big-endian magnitude
        for (i in y.indices) {
            val idx = y.size - 1 - i // least-significant first
            if (i < len) out[i] = y[idx]
        }
        if (point.isXOdd) out[len - 1] = (out[len - 1].toInt() or 0x80).toByte()
        return out
    }

    /** Rebuild a SunEC [PublicKey] from RAW RFC-8032 bytes (inverse of [rawPublicKey]). */
    fun publicKeyFromRaw(raw: ByteArray, curve: Curve): PublicKey {
        val le = raw.copyOf()
        val last = le.size - 1
        val xOdd = (le[last].toInt() and 0x80) != 0
        le[last] = (le[last].toInt() and 0x7f).toByte()
        val be = ByteArray(le.size) { le[le.size - 1 - it] }
        val y = BigInteger(1, be)
        val point = EdECPoint(xOdd, y)
        return KeyFactory.getInstance(algo(curve))
            .generatePublic(EdECPublicKeySpec(spec(curve), point))
    }
}
